go/tools/builders/stdliblist.go

```diff
@@ -0,0 +1,209 @@
+	// Dependency information
+	Imports   []string          // import paths used by this package
+	ImportMap map[string]string // map from source import to ImportPath (identity entries omitted)
+	// Error information
+	Incomplete bool                 // this package or a dependency has an error
+	Error      *flatPackagesError   // error loading package
+	DepsErrors []*flatPackagesError // errors loading dependencies
+}
+
+func stdlibPackageID(importPath string) string {
+	return "@io_bazel_rules_go//stdlib:" + importPath
+}
+
+func execRootPath(execRoot, p string) string {
+	dir, _ := filepath.Rel(execRoot, p)
+	return filepath.Join("__BAZEL_EXECROOT__", dir)
+}
+
+func absoluteSourcesPaths(execRoot, pkgDir string, srcs []string) []string {
+	ret := []string{}
+	pkgDir = execRootPath(execRoot, pkgDir)
+	for _, src := range srcs {
+		ret = append(ret, filepath.Join(pkgDir, src))
+	}
+	return ret
+}
+
+func packageToPackage(execRoot string, pkg *goListPackage) *flatPackage {
+	// Don't use generated files from the stdlib
+	goFiles := absoluteSourcesPaths(execRoot, pkg.Dir, pkg.GoFiles)
+
+	newPkg := &flatPackage{
+		ID:              stdlibPackageID(pkg.ImportPath),
+		Name:            pkg.Name,
+		PkgPath:         pkg.ImportPath,
+		ExportFile:      execRootPath(execRoot, pkg.Target),
+		Imports:         map[string]string{},
+		Standard:        pkg.Standard,
+		GoFiles:         goFiles,
+		CompiledGoFiles: goFiles,
+	}
+	for _, imp := range pkg.Imports {
+		newPkg.Imports[imp] = stdlibPackageID(imp)
+	}
+	// We don't support CGo for now
+	delete(newPkg.Imports, "C")
+	return newPkg
+}
+
+// stdlib builds the standard library in the appropriate mode into a new goroot.
+func stdliblist(args []string) error {
+	// process the args
+	flags := flag.NewFlagSet("stdliblist", flag.ExitOnError)
+	goenv := envFlags(flags)
+	out := flags.String("out", "", "Path to output go list json")
+	if err := flags.Parse(args); err != nil {
+		return err
+	}
+	if err := goenv.checkFlags(); err != nil {
+		return err
+	}
+
+	// Ensure paths are absolute.
+	absPaths := []string{}
+	for _, path := range filepath.SplitList(os.Getenv("PATH")) {
+		absPaths = append(absPaths, abs(path))
+	}
+	os.Setenv("PATH", strings.Join(absPaths, string(os.PathListSeparator)))
+	os.Setenv("GOROOT", abs(os.Getenv("GOROOT")))
+	// Make sure we have an absolute path to the C compiler.
+	// TODO(#1357): also take absolute paths of includes and other paths in flags.
+	os.Setenv("CC", abs(os.Getenv("CC")))
+
+	execRoot := abs(".")
+
+	cachePath := abs(*out + ".gocache")
+	defer os.RemoveAll(cachePath)
+	os.Setenv("GOCACHE", cachePath)
+	os.Setenv("GOPATH", cachePath)
+
+	listArgs := goenv.goCmd("list")
+	if len(build.Default.BuildTags) > 0 {
+		listArgs = append(listArgs, "-tags", strings.Join(build.Default.BuildTags, " "))
+	}
+	listArgs = append(listArgs, "-json", "builtin", "std", "runtime/cgo")
+
+	jsonFile, err := os.Create(*out)
+	if err != nil {
+		return err
+	}
+	defer jsonFile.Close()
```

Comment:
Check for an error when closing writable files:

```
defer func() {
  if cerr := jsonFile.Close(); err == nil && cerr != nil {
    err = cerr
  }
}()
```

Reply:
I mean, yes, but is there a cleaner way to do this ?